### PR TITLE
Shyrma concat

### PR DESCRIPTION
**Pull request by shyrma**
- provide deconv2d/3d (ff and bp) based on corresponding mkl dnn api
- add possibility to pass axis as last input array in concat op
- corrcect sumation in bias_add_bp op for NHWC case

### CODE REVIEW

File: libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp [hunk 1]
```diff
@@ -58,8 +58,8 @@ CUSTOM_OP_IMPL(deconv2d, 2, 1, false, 0, 9) {
     int indIOioC, indIiH, indWoC, indWiC, indWkH, indOoH;       // corresponding indexes
     ConvolutionUtils::getSizesAndIndexesConv2d(isNCHW, *input, *output, bS, iC, iH, iW, oC, oH, oW, indIOioC, indIiH, indWoC, indWiC, indWkH, indOoH);
 
-    std::string expectedWeightsShape = ShapeUtils::shapeAsString({kH, kW, oC, iC});
-    REQUIRE_TRUE(expectedWeightsShape == ShapeUtils::shapeAsString(weights), 0, "CUSTOM DECONV2D OP: wrong shape of weights array, expected is %s, but got %s instead !", expectedWeightsShape.c_str(), ShapeUtils::shapeAsString(weights).c_str());
+    std::vector<Nd4jLong> expectedWeightsShape = {kH, kW, oC, iC};
+    REQUIRE_TRUE(expectedWeightsShape == weights->getShapeAsVector(), 0, "CUSTOM DECONV2D OP: wrong shape of weights array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedWeightsShape).c_str(), ShapeUtils::shapeAsString(weights).c_str());
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp]: Why are we doing creating 2 vectors here? Can't we make a method, that will compare EXISTING shapeInfo to `expectedWeightsShape`?

File: libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp [hunk 2]
```diff
@@ -131,8 +131,8 @@ DECLARE_SHAPE_FN(deconv2d) {
     const int iC = inputShapeInfo[indIOioC+1];                   // input channels
     const int oC = weightsShapeInfo[indWoC+1];                   // output channels
 
-    std::string expectedWeightsShape = ShapeUtils::shapeAsString({kH, kW, oC, iC});
-    REQUIRE_TRUE(expectedWeightsShape == ShapeUtils::shapeAsString(weightsShapeInfo), 0, "CUSTOM DECONV2D OP: wrong shape of weights array, expected is %s, but got %s instead !", expectedWeightsShape.c_str(), ShapeUtils::shapeAsString(weightsShapeInfo).c_str());
+    std::vector<Nd4jLong> expectedWeightsShape = {kH, kW, oC, iC};
+    REQUIRE_TRUE(expectedWeightsShape == ShapeUtils::shapeAsVector(weightsShapeInfo), 0, "CUSTOM DECONV2D OP: wrong shape of weights array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedWeightsShape).c_str(), ShapeUtils::shapeAsString(weightsShapeInfo).c_str());
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp]: Same as above

File: libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp [hunk 3]
```diff
@@ -196,15 +196,18 @@ CUSTOM_OP_IMPL(deconv2d_bp, 3, 2, false, 0, 9) {
...
+    std::vector<Nd4jLong> expectedWeightsShape = {kH, kW, oC, iC};
+    REQUIRE_TRUE(expectedGradOShape == gradO->getShapeAsVector(), 0,  "CUSTOM DECONV2D_BP OP: wrong shape of output gradients (next epsilon) array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedGradOShape).c_str(), ShapeUtils::shapeAsString(gradO).c_str());
+    REQUIRE_TRUE(expectedWeightsShape == weights->getShapeAsVector(), 0, "CUSTOM DECONV2D_BP OP: wrong shape of weights array, expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedWeightsShape).c_str(), ShapeUtils::shapeAsString(weights).c_str());
     if(bias)
         REQUIRE_TRUE(bias->rankOf() <= 2 && oC == bias->lengthOf(), 0, "CUSTOM DECONV2D_BP OP: wrong shape of array with biases, expected rank, length: <=2, %i, but got %i, %i instead !", oC, bias->rankOf(), bias->lengthOf());
 
-    if(isSameMode)                       // SAME
-        ConvolutionUtils::calcPadding2D(pH, pW, oH, oW, iH, iW, kH, kW, sH, sW, dH, dW);
+    if(isSameMode){                       // SAME
+        //Note: we're intentionally swapping iH and oH, to calculated the padding for a"normal" conv (not deconv) forward pass
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/nn/convo/deconv2d.cpp:207]: 👍

File: libnd4j/include/ops/declarable/generic/nn/convo/deconv2d_tf.cpp [hunk 4]
```diff
@@ -52,26 +52,26 @@ CUSTOM_OP_IMPL(deconv2d_tf, 3, 1, false, 0, 9) {
...
-    std::string expectedWeightsShape = ShapeUtils::shapeAsString({kH, kW, iC, oC});
-    REQUIRE_TRUE(expectedGradOShape == ShapeUtils::shapeAsString(gradO), 0,  "CUSTOM DECONV2D_TF OP: wrong shape of input array, basing on array with output shape expected is %s, but got %s instead !", expectedGradOShape.c_str(), ShapeUtils::shapeAsString(gradO).c_str());
-    REQUIRE_TRUE(expectedWeightsShape == ShapeUtils::shapeAsString(weights), 0, "CUSTOM DECONV2D_TF OP: wrong shape of weights array, expected is %s, but got %s instead !", expectedWeightsShape.c_str(), ShapeUtils::shapeAsString(weights).c_str());
+    std::vector<Nd4jLong> expectedGradOShape   = ShapeUtils::composeShapeUsingDimsAndIdx({bS,oC,trueoH,trueoW,  0,indIOioC,indOoH,indOoH+1});
+    std::vector<Nd4jLong> expectedWeightsShape = {kH, kW, iC, oC};
+    REQUIRE_TRUE(expectedGradOShape == gradO->getShapeAsVector(), 0,  "CUSTOM DECONV2D_TF OP: wrong shape of input array, basing on array with output shape expected is %s, but got %s instead !", ShapeUtils::shapeAsString(expectedGradOShape).c_str(), ShapeUtils::shapeAsString(gradO).c_str());
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/nn/convo/deconv2d_tf.cpp]: Same as above

File: libnd4j/include/ops/declarable/generic/nn/convo/deconv2d_tf.cpp [hunk 5]
```diff
@@ -88,11 +88,10 @@ DECLARE_SHAPE_FN(deconv2d_tf) {
     auto gradIShapeShapeInfo = inputShape->at(0);                                             // [4]
 
     const int rank = 4;
-    
+
     REQUIRE_TRUE(weightsShapeInfo[0]  == rank, 0, "CUSTOM DECONV2D_TF OP: rank of weights array must be equal to %i, but got %i instead !", rank, weightsShapeInfo[0]);
     REQUIRE_TRUE(gradOShapeInfo[0]    == rank, 0, "CUSTOM DECONV2D_TF OP: rank of input array must be equal to %i, but got %i instead !", rank, gradOShapeInfo[0]);
-    REQUIRE_TRUE(gradIShapeShapeInfo[0] == 1,    0, "CUSTOM DECONV2D_TF OP: rank of array with output shape must be equal to %i, but got %i instead !", 1, gradIShapeShapeInfo[0]);    
-    
+    REQUIRE_TRUE(gradIShapeShapeInfo[0] == 1,    0, "CUSTOM DECONV2D_TF OP: rank of array with output shape must be equal to %i, but got %i instead !", 1, gradIShapeShapeInfo[0]);
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/nn/convo/deconv2d_tf.cpp]: Please use shape::sizeAt() and shape::rank() instead. Don't use direct buffer comparison.

**Review by raver119 — CHANGES_REQUESTED** [review]
Few minor changes required. Looks great otherwise.

**Comment by shyrma** [comment]
thank for reasonable comments,
corrrected !

**Review by raver119 — APPROVED** [review]

**Pull request merged**